### PR TITLE
DOP-3130: Endpoints for `/repos/{repo_name}/versions`

### PR DESCRIPTION
**Pull request by drew-beckmen**
### Ticket

[DOP-3130](https://jira.mongodb.org/browse/DOP-3130)

### Description

This pull request introduces the GET, PUT, and POST endpoints for `/repos/{repo_name}/versions`. In addition, this PR:

- establishes a new pattern for validation by adding validation classes into a `validators` directory and using validations as a dependency. In general, the pattern is that the arguments to a path operation function become the return value of a dependency function specified in `dependencies.py`. Functions in `dependencies.py` can initialize a Pydantic model from the `validators` directory to carry out validations.
- fixes a bug for local development caused by cookies being preserved on `localhost`
- adds the `id` field to nested documents in the `branches` field
- ensures consistency of error response schemas (and document these schemas in the OpenAPI spec)

### CODE REVIEW

**Review by rayangler — COMMENTED** [review]
Mostly just nits and questions for now

File: backend/api/core/validators/group_validator.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        existing_groups: list[Group] = repo.groups
+
+        # Every version in included branches should be in the repo's array of branches
+        missing_versions = list(
+            set(new_group.included_branches) - cls.get_version_names(repo)
+        )
+        if len(missing_versions) > 0:
+            err_msg = (
+                f'Attempting to use version "{",".join(missing_versions)}" in {new_group_label}, '
+                f'but version "{",".join(missing_versions)}" does not exist.'
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply backend/api/core/validators/group_validator.py:36]: I like it!

File: backend/api/core/validators/reorder_validator.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ListT = TypeVar("ListT")
+
+
+class ReorderPayloadValidator(GenericModel, Generic[ListT], ReorderItemPayload):
+    reorderingList: list[ListT]
+    repo: Repo
+
+    @root_validator(pre=True)
+    def validate_indexes(cls, values):
+        print(ReorderItemPayload.__fields__.keys())
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply backend/api/core/validators/reorder_validator.py:20]: oops i should grep for these before I open a PR 😄 - left over from debugging

File: backend/api/core/validators/version_validator.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        alias_errors = []
+        new_aliases = new_version.url_aliases
+        if new_aliases is not None:
+            existing_aliases = filter(
+                None, [version.url_aliases for version in existing_versions]
+            )
+            for aliases in existing_aliases:
+                if set(new_version.url_aliases) & set(aliases):
+                    alias_errors.append(
+                        f"urlAliases: one of {new_aliases} is already in use"
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply backend/api/core/validators/version_validator.py:30]: agreed!

File: backend/api/core/validators/version_validator.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                )
+            )
+            if num_stable_branches > 0:
+                return "Only one stable branch can be present"
+
+    @classmethod
+    def validate_unique_version_selector(cls, new_version, existing_versions):
+        version_selector_label = new_version.version_selector_label
+        if version_selector_label is not None:
+            if version_selector_label in set(
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply backend/api/core/validators/version_validator.py:49]: good call - it's the same time complexity [O(n) for conversion vs O(n) for list search] and more space so not worth it if the set isn't used for anything else

File: backend/api/models/version.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    url_slug: str | None = Field(alias="urlSlug")
+    version_selector_label: str | None = Field(alias="versionSelectorLabel")
+    is_stable_branch: bool | None = Field(alias="isStableBranch")
+
+    # If validation on git_branch name fails, it won't be included in values.
+    # https://pydantic-docs.helpmanual.io/usage/validators/
+    @validator("url_slug", always=True)
+    def url_slug_validator(cls, v, values):
+        if "git_branch_name" not in values:
+            return v
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply backend/api/models/version.py:22]: `git_branch_name` is a required field, so Pydantic will respond with a validation error if it is None. However, Pydantic runs all validators before returning the exceptions, so if `git_branch_name=None` (i.e. it is missing or left blank), the rest of the field validations will run.

According to the docs, if a field is None, it's not passed through to `values`. Since `url_slug_validatoir` depends on `git_branch_name` being a key in `values`, we need to catch the case in which `git_branch_name=None`

The case when two versions have the same branch name would be in `VersionValidator`. The `Version` model is used for validations we can run without info on the rest of the repo.

File: backend/api/v1/routes/groups.py [hunk 6]
```diff
@@ -1,25 +1,66 @@
...
-@router.post(GROUPS_INDEX_PATH, response_model=Group)
-async def create_group(repo_name: str, group: Group):
-    await insert_new_group(repo_name, group)
-    return group
+@router.post(
+    GROUPS_INDEX_PATH,
+    response_model=Group,
+    description="Create a new group for a specific repo",
+)
+async def create_group(new_group_params: GroupValidator = Depends(new_group_validator)):
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply backend/api/v1/routes/groups.py:46]: I think this makes sense! I like the idea of `new_group_validator` being a dependency function whose responsibility is to return a `ValidatedGroup` (rather than just instantiating a `GroupValidator`

File: backend/api/v1/routes/groups.py [hunk 7]
```diff
@@ -1,25 +1,66 @@
...
+)
+async def reorder_group(
+    reorder_params: ReorderPayloadValidator = Depends(reorder_group_validator),
+):
+    updated_repo = await reorder_repo_group(
+        repo=reorder_params.repo,
+        current_index=reorder_params.current_index,
+        target_index=reorder_params.target_index,
+    )
+    return updated_repo
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply backend/api/v1/routes/groups.py:66]: I wasn't sure what we wanted the API response from these `PUT` operations to be. For versions, I was returning the whole repo and projecting it onto a `RepoVersionsView`, so I implemented the same pattern for groups. The reason I chose to return the list of versions/groups was that it's the only response payload that illustrates the new state of the data server side. For a `POST`, returning the created version/group reflects the data change server side. For a `PUT`, returning the moved version/group doesn't provide the client with information on the new state server-side (because each individual version/group doesn't contain information on its ordering).

tl;dr -> I through returning `versions: []` or `groups: []` would be helpful for the client. I also think the pattern makes sense for the `/groups` and `/versions` endpoints as we aren't a specific `/group` or `/version`

[the remark this answers is not in the document]
Comment (rayangler) [reply backend/api/v1/routes/groups.py:66]: I didn't see the response model being used! Returning the list makes sense.

File: backend/api/models/version.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rayangler) [reply backend/api/models/version.py:22]: Gotcha, so the purpose of this check here is to just prevent throwing a key error in the event that a `Version` is trying to be constructed with a missing `git_branch_name`. My assumption was that since the field is required, this validation automatically wouldn't be reached due to a prior exception, but this flow makes sense as well. Thanks!

File: backend/api/core/validators/valid_group.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            flattened_existing_branches = [
+                branch for branches in existing_included_branches for branch in branches
+            ]
+            previously_used_versions = list(
+                set(new_group.included_branches) & set(flattened_existing_branches)
+            )
+            if len(previously_used_versions) > 0:
+                err_msg = (
+                    f'Attempting to use version {",".join(previously_used_versions)} in {new_group_label}. '
+                    f'{",".join(previously_used_versions)} already exists in another group.'
```
Comment (rayangler) [inline backend/api/core/validators/valid_group.py:57]: (Nit) Similar concerns about whitespace after commas and message conciseness when multiple versions are printed out.

```
Versions {foo, bar, baz} are already used in another group.
```

**Review by rayangler — APPROVED** [review]
LGTM with one non-blocking nit

File: backend/api/core/validators/valid_group.py (hunk 8 above)
Comment (drew-beckmen) [reply backend/api/core/validators/valid_group.py:57]: will fix!
